tests/browser_test.py

```diff
@@ -36,86 +67,407 @@ class TestBrowserFunctions(unittest.TestCase):
+        r = self.browser.open_url(BASE_URL + '/html_test.html')
+        self.assertRaises(LinkNotFound, self.browser.follow_link, '#bad-link')
+
+        # Good link
+        self.browser.open_url(BASE_URL + '/html_test.html')
+        r = self.browser.follow_link('#test_link')
+        self.assertEqual(r.status_code, 200)
+        self.assertEqual(r.url, BASE_URL + '/basic_page.html')
+
+        # Good link + regex on href
+        self.browser.open_url(BASE_URL + '/html_test.html')
+        r = self.browser.follow_link('#test_link', url_regex='.*\.html')
+        self.assertEqual(r.status_code, 200)
+        self.assertEqual(r.url, BASE_URL + '/basic_page.html')
+
+        # Good link + regex on anchor inner html
+        self.browser.open_url(BASE_URL + '/html_test.html')
+        r = self.browser.follow_link('#test_link', url_regex='.*Basic.*')
+        self.assertEqual(r.status_code, 200)
+        self.assertEqual(r.url, BASE_URL + '/basic_page.html')
+
+        # Good link + bad regex
+        r = self.browser.open_url(BASE_URL + '/html_test.html')
+        self.assertRaises(
+            LinkNotFound,
+            self.browser.follow_link,
+            '#test_link',
+            url_regex='.*\.php')
+
+        # Browser.clean_session
+        # Verify browser is not clean
+        self.assertIsNotNone(self.browser._url)
+        self.assertIsNotNone(self.browser._html)
+        self.assertIsNotNone(self.browser._response)
+        self.browser.add_header('special', 'value')
+        if self.browser.history_object is not None:
+            self.browser.history_object.history.append('something')
+            self.assertGreater(len(self.browser.history_object.history), 0)
+        # Do browser clean
+        self.browser.clean_browser()
+        # Verify results
+        self.assertEqual(self.browser._base_url, BASE_URL)
+        self.assertIsNone(self.browser._url)
+        self.assertIsNone(self.browser._html)
+        self.assertIsNone(self.browser._response)
+        self.assertNotIn('special', self.browser.session.headers)
+        if self.browser.history_object is not None:
+            self.assertIsNone(self.browser.history)
+
+        # Browser._process_response
+        # Handle requests responses
+        self.browser.clean_browser()
+        self.assertIsNone(self.browser._url)
+        self.assertIsNone(self.browser._html)
+        self.assertIsNone(self.browser._response)
+        r = self.browser.session.get(BASE_URL + '/html_test.html')
+        r_html = self.browser._process_response(r)
+        self.assertEqual(r.url, self.browser._url)
+        self.assertIsNotNone(self.browser._html)
+        self.assertEqual(self.browser._response, r_html)
+
+        # Handle urllib responses
+        self.browser.clean_browser()
+        self.assertIsNone(self.browser._url)
+        self.assertIsNone(self.browser._html)
+        self.assertIsNone(self.browser._response)
+        r = urllib.urlopen(BASE_URL + '/html_test.html')
```

Comment:
in python3 urllib doesn't have an urlopen method anymore, but you can use : ``r = urllib.request.urlopen(BASE_URL + '/html_test.html')`` instead